codespan/src/file.rs

```diff
@@ -0,0 +1,183 @@
+                    .source
+                    .get(line_start_index..byte_index)
+                    .ok_or_else(|| {
+                        let given = byte_index as usize;
+                        if given >= self.source_span().end() {
+                            let span = self.source_span();
+                            FileError::OutOfBounds { given, span }
+                        } else {
+                            FileError::InvalidCharBoundary { given }
+                        }
+                    })?;
+
+                Ok(Position::new(line_index, line_src.len()))
+            }
+        }
+    }
+
+    fn source_slice(&self, span: Span) -> Result<&str> {
+        let start = span.start;
+        let end = span.end;
+
+        self.source.get(start..end).ok_or_else(|| {
+            // tracing::error!("Failed to create source span");
+            panic!()
+        })
+    }
+
+    fn source_span(&self) -> Span {
+        Span::new(0, self.source.len())
+    }
+
+    pub fn position_to_byte_index(&self, position: Position) -> Result<usize> {
+        let line_span = &self.line_span(position.line)?;
+        let byte_offset = position.character;
+
+        Ok(line_span.start() + byte_offset)
+    }
+
+    pub fn range_to_byte_span(
+        &self,
+        range: &Range,
+    ) -> Result<std::ops::Range<usize>> {
+        Ok(self.position_to_byte_index(range.start)?
+            ..self.position_to_byte_index(range.end)?)
+    }
+
+    pub fn byte_index_to_position(&self, byte_index: usize) -> Result<Position> {
+        let location = self.location(byte_index)?;
+
+        Ok(location)
+    }
+
+    pub fn byte_span_to_range(&self, span: Span) -> Result<Range> {
+        Ok(Range {
+            start: self.byte_index_to_position(span.start())?,
+            end: self.byte_index_to_position(span.end())?,
+        })
+    }
+
+    pub fn get_line(&self, line_index: usize) -> Result<Span> {
+        self.line_span(line_index)
+    }
+
+    #[allow(dead_code)]
+    pub fn get_line_source(&self, span: Span) -> Result<&str> {
+        self.source_slice(span)
+    }
+
+    pub fn get_word_at_position(&self, position: Position) -> Result<&str> {
+        let line = self.source_slice(self.line_span(position.line)?)?;
+        let range = find_word_at_pos(line, position.character);
+
+        Ok(line.get(range.0..range.1).unwrap())
+    }
+}
+
+pub fn find_word_at_pos(line: &str, col: usize) -> (usize, usize) {
+    let line_ = format!("{} ", line);
+    let is_ident_char = |c: char| c.is_alphanumeric() || c == '_' || c == '@';
+
+    let start = line_
+        .chars()
+        .enumerate()
+        .take(col)
+        .filter(|&(_, c)| !is_ident_char(c))
+        .last()
+        .map(|(i, _)| i + 1)
+        .unwrap_or(0);
+
+    let mut end = line_
+        .chars()
+        .enumerate()
+        .skip(col)
+        .filter(|&(_, c)| !is_ident_char(c));
+
+    let end = end.next();
+    (start, end.map(|(i, _)| i).unwrap_or(col))
```

Comment:
Why not just add the `next()` and `map()` onto the chain?
```suggestion
    let mut end = line_
        .chars()
        .enumerate()
        .skip(col)
        .filter(|&(_, c)| !is_ident_char(c))
        .next()
        .map(|(i, _)| i);

    (start, end).unwrap_or(col))
```